#include "MainWindow.h"
#include "ui_MainWindow.h"
#include "MachineTableModel.h"
#include "../database/FleetDatabase.h"
#include <QTableView>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QComboBox>
#include <QGroupBox>
#include <QHeaderView>
#include <QMessageBox>
#include <QDebug>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , m_tableModel(nullptr)
    , m_tableView(nullptr)
    , m_statusFilter(nullptr)
{
    ui->setupUi(this);
    setupUI();
    connectSignals();
    
    // –ó–∞–≥—Ä—É–∑–∫–∞ –¥–∞–Ω–Ω—ã—Ö
    m_tableModel->loadData();
    updateStatusBar();
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::setupUI()
{
    setupSidebar();
    setupTable();
    setupDetailsPanel();
}

void MainWindow::setupSidebar()
{
    // –ë–æ–∫–æ–≤–∞—è –ø–∞–Ω–µ–ª—å —Å –Ω–∞–≤–∏–≥–∞—Ü–∏–µ–π –∏ —Ñ–∏–ª—å—Ç—Ä–∞–º–∏
    QVBoxLayout *sidebarLayout = new QVBoxLayout(ui->sidebarWidget);
    sidebarLayout->setContentsMargins(8, 8, 8, 8);
    sidebarLayout->setSpacing(8);
    
    // –ó–∞–≥–æ–ª–æ–≤–æ–∫ –Ω–∞–≤–∏–≥–∞—Ü–∏–∏
    QLabel *navTitle = new QLabel("–ù–ê–í–ò–ì–ê–¶–ò–Ø");
    navTitle->setStyleSheet("color: #858585; font-weight: bold; font-size: 11px; padding-bottom: 4px;");
    sidebarLayout->addWidget(navTitle);
    
    // –ö–Ω–æ–ø–∫–∏ –Ω–∞–≤–∏–≥–∞—Ü–∏–∏
    QPushButton *btnFleet = new QPushButton("üìã –ü–∞—Ä–∫ —Ç–µ—Ö–Ω–∏–∫–∏");
    btnFleet->setStyleSheet(R"(
        QPushButton {
            text-align: left;
            padding: 8px 12px;
            background-color: #094771;
            color: white;
            border: none;
            border-radius: 2px;
        }
        QPushButton:hover {
            background-color: #0e639c;
        }
    )");
    sidebarLayout->addWidget(btnFleet);
    
    QPushButton *btnProjects = new QPushButton("üèóÔ∏è –ü—Ä–æ–µ–∫—Ç—ã");
    btnProjects->setStyleSheet(R"(
        QPushButton {
            text-align: left;
            padding: 8px 12px;
            background-color: transparent;
            color: #cccccc;
            border: none;
        }
        QPushButton:hover {
            background-color: #2a2d2e;
        }
    )");
    sidebarLayout->addWidget(btnProjects);
    
    QPushButton *btnHistory = new QPushButton("üìú –ò—Å—Ç–æ—Ä–∏—è");
    btnHistory->setStyleSheet(btnProjects->styleSheet());
    sidebarLayout->addWidget(btnHistory);
    
    QPushButton *btnReports = new QPushButton("üìä –û—Ç—á—ë—Ç—ã");
    btnReports->setStyleSheet(btnProjects->styleSheet());
    sidebarLayout->addWidget(btnReports);
    
    QPushButton *btnSettings = new QPushButton("‚öôÔ∏è –ù–∞—Å—Ç—Ä–æ–π–∫–∏");
    btnSettings->setStyleSheet(btnProjects->styleSheet());
    sidebarLayout->addWidget(btnSettings);
    
    sidebarLayout->addSpacing(20);
    
    // –§–∏–ª—å—Ç—Ä –ø–æ —Å—Ç–∞—Ç—É—Å—É
    QLabel *filterTitle = new QLabel("–§–ò–õ–¨–¢–†");
    filterTitle->setStyleSheet("color: #858585; font-weight: bold; font-size: 11px; padding-bottom: 4px;");
    sidebarLayout->addWidget(filterTitle);
    
    m_statusFilter = new QComboBox();
    m_statusFilter->addItem("–í—Å–µ —Å—Ç–∞—Ç—É—Å—ã");
    m_statusFilter->addItem("–°–≤–æ–±–æ–¥–Ω–∞");
    m_statusFilter->addItem("–ù–∞ –æ–±—ä–µ–∫—Ç–µ");
    m_statusFilter->addItem("–í —Ä–µ–º–æ–Ω—Ç–µ");
    m_statusFilter->addItem("–°–ø–∏—Å–∞–Ω–∞");
    m_statusFilter->setStyleSheet(R"(
        QComboBox {
            padding: 6px;
            background-color: #3c3c3c;
            color: #cccccc;
            border: 1px solid #555555;
            border-radius: 2px;
        }
        QComboBox:hover {
            background-color: #4a4a4a;
        }
        QComboBox::drop-down {
            border: none;
        }
        QComboBox QAbstractItemView {
            background-color: #3c3c3c;
            color: #cccccc;
            selection-background-color: #094771;
        }
    )");
    sidebarLayout->addWidget(m_statusFilter);
    
    sidebarLayout->addStretch();
}

void MainWindow::setupTable()
{
    // –°–æ–∑–¥–∞—ë–º –º–æ–¥–µ–ª—å —Ç–∞–±–ª–∏—Ü—ã
    m_tableModel = new MachineTableModel(this);
    
    // –°–æ–∑–¥–∞—ë–º –ø—Ä–µ–¥—Å—Ç–∞–≤–ª–µ–Ω–∏–µ —Ç–∞–±–ª–∏—Ü—ã
    m_tableView = new QTableView();
    m_tableView->setModel(m_tableModel);
    m_tableView->setSelectionBehavior(QAbstractItemView::SelectRows);
    m_tableView->setSelectionMode(QAbstractItemView::SingleSelection);
    m_tableView->setAlternatingRowColors(true);
    m_tableView->verticalHeader()->setVisible(false);
    m_tableView->horizontalHeader()->setStretchLastSection(true);
    m_tableView->setStyleSheet(R"(
        QTableView {
            background-color: #1e1e1e;
            color: #d4d4d4;
            gridline-color: #2d2d2d;
            border: none;
            selection-background-color: #264f78;
        }
        QTableView::item:selected {
            background-color: #264f78;
        }
        QTableView::item:hover {
            background-color: #2a2d2e;
        }
        QHeaderView::section {
            background-color: #2d2d2d;
            color: #cccccc;
            padding: 6px;
            border: none;
            border-bottom: 1px solid #3e3e3e;
            font-weight: bold;
        }
    )");
    
    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —à–∏—Ä–∏–Ω—É –∫–æ–ª–æ–Ω–æ–∫
    m_tableView->setColumnWidth(0, 200); // –ù–∞–∑–≤–∞–Ω–∏–µ
    m_tableView->setColumnWidth(1, 150); // –¢–∏–ø
    m_tableView->setColumnWidth(2, 150); // –°–µ—Ä–∏–π–Ω—ã–π –Ω–æ–º–µ—Ä
    m_tableView->setColumnWidth(3, 100); // –ì–æ–¥
    m_tableView->setColumnWidth(4, 120); // –°—Ç–∞—Ç—É—Å
    m_tableView->setColumnWidth(5, 120); // –°—Ç–æ–∏–º–æ—Å—Ç—å
    
    // –î–æ–±–∞–≤–ª—è–µ–º —Ç–∞–±–ª–∏—Ü—É –≤ –∫–æ–Ω—Ç–µ–π–Ω–µ—Ä
    QVBoxLayout *tableLayout = new QVBoxLayout(ui->tableContainer);
    tableLayout->setContentsMargins(0, 0, 0, 0);
    tableLayout->addWidget(m_tableView);
}

void MainWindow::setupDetailsPanel()
{
    m_detailsPanel = ui->detailsContainer;
    QVBoxLayout *layout = new QVBoxLayout(m_detailsPanel);
    layout->setContentsMargins(12, 12, 12, 12);
    layout->setSpacing(10);
    
    // –ó–∞–≥–æ–ª–æ–≤–æ–∫ –ø–∞–Ω–µ–ª–∏
    QLabel *title = new QLabel("–î–ï–¢–ê–õ–ò –¢–ï–•–ù–ò–ö–ò");
    title->setStyleSheet("color: #858585; font-weight: bold; font-size: 11px; padding-bottom: 4px;");
    layout->addWidget(title);
    
    // –ì—Ä—É–ø–ø–∞ —Å –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–µ–π
    QGroupBox *infoGroup = new QGroupBox();
    infoGroup->setStyleSheet(R"(
        QGroupBox {
            background-color: #252526;
            border: 1px solid #3e3e3e;
            border-radius: 4px;
            padding: 12px;
        }
    )");
    QVBoxLayout *infoLayout = new QVBoxLayout(infoGroup);
    infoLayout->setSpacing(8);
    
    // –°–æ–∑–¥–∞—ë–º –º–µ—Ç–∫–∏ –¥–ª—è –æ—Ç–æ–±—Ä–∞–∂–µ–Ω–∏—è –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–∏
    auto createInfoLabel = [](const QString& title, QLabel** valueLabel) -> QWidget* {
        QWidget *widget = new QWidget();
        QVBoxLayout *vbox = new QVBoxLayout(widget);
        vbox->setContentsMargins(0, 0, 0, 0);
        vbox->setSpacing(2);
        
        QLabel *titleLabel = new QLabel(title);
        titleLabel->setStyleSheet("color: #858585; font-size: 10px;");
        vbox->addWidget(titleLabel);
        
        *valueLabel = new QLabel("‚Äî");
        (*valueLabel)->setStyleSheet("color: #d4d4d4; font-size: 12px; font-weight: 500;");
        (*valueLabel)->setWordWrap(true);
        vbox->addWidget(*valueLabel);
        
        return widget;
    };
    
    infoLayout->addWidget(createInfoLabel("–ù–ê–ó–í–ê–ù–ò–ï", &m_detailsName));
    infoLayout->addWidget(createInfoLabel("–¢–ò–ü –¢–ï–•–ù–ò–ö–ò", &m_detailsType));
    infoLayout->addWidget(createInfoLabel("–°–ï–†–ò–ô–ù–´–ô –ù–û–ú–ï–†", &m_detailsSerial));
    infoLayout->addWidget(createInfoLabel("–ì–û–î –í–´–ü–£–°–ö–ê", &m_detailsYear));
    infoLayout->addWidget(createInfoLabel("–°–¢–ê–¢–£–°", &m_detailsStatus));
    infoLayout->addWidget(createInfoLabel("–°–¢–û–ò–ú–û–°–¢–¨", &m_detailsCost));
    infoLayout->addWidget(createInfoLabel("–¢–ï–ö–£–©–ò–ô –ü–†–û–ï–ö–¢", &m_detailsProject));
    infoLayout->addWidget(createInfoLabel("–ù–ê–ó–ù–ê–ß–ï–ù –°", &m_detailsAssignedDate));
    
    layout->addWidget(infoGroup);
    layout->addStretch();
    
    // –ò–∑–Ω–∞—á–∞–ª—å–Ω–æ –ø–æ–∫–∞–∑—ã–≤–∞–µ–º –ø—É—Å—Ç—É—é –ø–∞–Ω–µ–ª—å
    updateDetailsPanel(nullptr);
}

void MainWindow::connectSignals()
{
    // –ü–æ–¥–∫–ª—é—á–∞–µ–º –¥–µ–π—Å—Ç–≤–∏—è –º–µ–Ω—é –∏ toolbar
    connect(ui->actionAdd, &QAction::triggered, this, &MainWindow::onAddMachine);
    connect(ui->actionEdit, &QAction::triggered, this, &MainWindow::onEditMachine);
    connect(ui->actionDelete, &QAction::triggered, this, &MainWindow::onDeleteMachine);
    connect(ui->actionAssignToProject, &QAction::triggered, this, &MainWindow::onAssignToProject);
    connect(ui->actionReturnFromProject, &QAction::triggered, this, &MainWindow::onReturnFromProject);
    connect(ui->actionSendToRepair, &QAction::triggered, this, &MainWindow::onSendToRepair);
    connect(ui->actionAbout, &QAction::triggered, this, &MainWindow::onAbout);
    connect(ui->actionExit, &QAction::triggered, this, &MainWindow::onExit);
    
    // –ü–æ–¥–∫–ª—é—á–∞–µ–º –≤—ã–±–æ—Ä —Å—Ç—Ä–æ–∫–∏ –≤ —Ç–∞–±–ª–∏—Ü–µ
    connect(m_tableView->selectionModel(), &QItemSelectionModel::selectionChanged,
            this, &MainWindow::onTableSelectionChanged);
    
    // –ü–æ–¥–∫–ª—é—á–∞–µ–º —Ñ–∏–ª—å—Ç—Ä –ø–æ —Å—Ç–∞—Ç—É—Å—É
    connect(m_statusFilter, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &MainWindow::onStatusFilterChanged);
}

void MainWindow::onAddMachine()
{
    // TODO: –°–æ–∑–¥–∞—Ç—å –¥–∏–∞–ª–æ–≥ –¥–æ–±–∞–≤–ª–µ–Ω–∏—è —Ç–µ—Ö–Ω–∏–∫–∏
    QMessageBox::information(this, "–î–æ–±–∞–≤–ª–µ–Ω–∏–µ", "–§—É–Ω–∫—Ü–∏—è –¥–æ–±–∞–≤–ª–µ–Ω–∏—è —Ç–µ—Ö–Ω–∏–∫–∏ –±—É–¥–µ—Ç —Ä–µ–∞–ª–∏–∑–æ–≤–∞–Ω–∞ –≤ —Å–ª–µ–¥—É—é—â–µ–π –≤–µ—Ä—Å–∏–∏");
}

void MainWindow::onEditMachine()
{
    auto machine = getSelectedMachine();
    if (!machine) {
        QMessageBox::warning(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ", "–í—ã–±–µ—Ä–∏—Ç–µ —Ç–µ—Ö–Ω–∏–∫—É –¥–ª—è —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏—è");
        return;
    }
    
    // TODO: –°–æ–∑–¥–∞—Ç—å –¥–∏–∞–ª–æ–≥ —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏—è
    QMessageBox::information(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ", 
                            QString("–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ: %1\n–§—É–Ω–∫—Ü–∏—è –±—É–¥–µ—Ç —Ä–µ–∞–ª–∏–∑–æ–≤–∞–Ω–∞ –≤ —Å–ª–µ–¥—É—é—â–µ–π –≤–µ—Ä—Å–∏–∏")
                            .arg(machine->getName()));
}

void MainWindow::onDeleteMachine()
{
    auto machine = getSelectedMachine();
    if (!machine) {
        QMessageBox::warning(this, "–£–¥–∞–ª–µ–Ω–∏–µ", "–í—ã–±–µ—Ä–∏—Ç–µ —Ç–µ—Ö–Ω–∏–∫—É –¥–ª—è —É–¥–∞–ª–µ–Ω–∏—è");
        return;
    }
    
    // –ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ —É–¥–∞–ª–µ–Ω–∏—è
    auto reply = QMessageBox::question(this, "–ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ —É–¥–∞–ª–µ–Ω–∏—è",
                                      QString("–£–¥–∞–ª–∏—Ç—å —Ç–µ—Ö–Ω–∏–∫—É \"%1\"?\n–≠—Ç–æ –¥–µ–π—Å—Ç–≤–∏–µ –Ω–µ–ª—å–∑—è –æ—Ç–º–µ–Ω–∏—Ç—å.")
                                      .arg(machine->getName()),
                                      QMessageBox::Yes | QMessageBox::No);
    
    if (reply == QMessageBox::Yes) {
        if (FleetDatabase::instance().deleteMachine(machine->getId())) {
            m_tableModel->loadData();
            updateStatusBar();
            updateDetailsPanel(nullptr);
            QMessageBox::information(this, "–£–¥–∞–ª–µ–Ω–∏–µ", "–¢–µ—Ö–Ω–∏–∫–∞ —É—Å–ø–µ—à–Ω–æ —É–¥–∞–ª–µ–Ω–∞");
        } else {
            QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å —Ç–µ—Ö–Ω–∏–∫—É");
        }
    }
}

void MainWindow::onAssignToProject()
{
    auto machine = getSelectedMachine();
    if (!machine) {
        QMessageBox::warning(this, "–ù–∞–∑–Ω–∞—á–µ–Ω–∏–µ –Ω–∞ –ø—Ä–æ–µ–∫—Ç", "–í—ã–±–µ—Ä–∏—Ç–µ —Ç–µ—Ö–Ω–∏–∫—É");
        return;
    }
    
    if (machine->getStatus() != MachineStatus::Available) {
        QMessageBox::warning(this, "–ù–∞–∑–Ω–∞—á–µ–Ω–∏–µ –Ω–∞ –ø—Ä–æ–µ–∫—Ç", 
                           "–ú–æ–∂–Ω–æ –Ω–∞–∑–Ω–∞—á–∞—Ç—å —Ç–æ–ª—å–∫–æ —Å–≤–æ–±–æ–¥–Ω—É—é —Ç–µ—Ö–Ω–∏–∫—É");
        return;
    }
    
    // TODO: –î–∏–∞–ª–æ–≥ –≤—ã–±–æ—Ä–∞ –ø—Ä–æ–µ–∫—Ç–∞
    QMessageBox::information(this, "–ù–∞–∑–Ω–∞—á–µ–Ω–∏–µ –Ω–∞ –ø—Ä–æ–µ–∫—Ç", 
                           "–§—É–Ω–∫—Ü–∏—è –Ω–∞–∑–Ω–∞—á–µ–Ω–∏—è –Ω–∞ –ø—Ä–æ–µ–∫—Ç –±—É–¥–µ—Ç —Ä–µ–∞–ª–∏–∑–æ–≤–∞–Ω–∞ –≤ —Å–ª–µ–¥—É—é—â–µ–π –≤–µ—Ä—Å–∏–∏");
}

void MainWindow::onReturnFromProject()
{
    auto machine = getSelectedMachine();
    if (!machine) {
        QMessageBox::warning(this, "–í–æ–∑–≤—Ä–∞—Ç —Å –ø—Ä–æ–µ–∫—Ç–∞", "–í—ã–±–µ—Ä–∏—Ç–µ —Ç–µ—Ö–Ω–∏–∫—É");
        return;
    }
    
    if (machine->getStatus() != MachineStatus::OnSite) {
        QMessageBox::warning(this, "–í–æ–∑–≤—Ä–∞—Ç —Å –ø—Ä–æ–µ–∫—Ç–∞", 
                           "–ú–æ–∂–Ω–æ –≤–µ—Ä–Ω—É—Ç—å —Ç–æ–ª—å–∫–æ —Ç–µ—Ö–Ω–∏–∫—É, –∫–æ—Ç–æ—Ä–∞—è –Ω–∞—Ö–æ–¥–∏—Ç—Å—è –Ω–∞ –æ–±—ä–µ–∫—Ç–µ");
        return;
    }
    
    // –û–±–Ω–æ–≤–ª—è–µ–º —Å—Ç–∞—Ç—É—Å
    machine->setStatus(MachineStatus::Available);
    machine->setCurrentProject("");
    machine->setAssignedDate(QDate());
    
    if (FleetDatabase::instance().updateMachine(machine)) {
        m_tableModel->loadData();
        updateStatusBar();
        onTableSelectionChanged(); // –û–±–Ω–æ–≤–ª—è–µ–º –ø–∞–Ω–µ–ª—å –¥–µ—Ç–∞–ª–µ–π
        QMessageBox::information(this, "–í–æ–∑–≤—Ä–∞—Ç —Å –ø—Ä–æ–µ–∫—Ç–∞", 
                               QString("–¢–µ—Ö–Ω–∏–∫–∞ \"%1\" –≤–æ–∑–≤—Ä–∞—â–µ–Ω–∞ –≤ –ø–∞—Ä–∫").arg(machine->getName()));
    } else {
        QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –æ–±–Ω–æ–≤–∏—Ç—å —Å—Ç–∞—Ç—É—Å —Ç–µ—Ö–Ω–∏–∫–∏");
    }
}

void MainWindow::onSendToRepair()
{
    auto machine = getSelectedMachine();
    if (!machine) {
        QMessageBox::warning(this, "–û—Ç–ø—Ä–∞–≤–∫–∞ –≤ —Ä–µ–º–æ–Ω—Ç", "–í—ã–±–µ—Ä–∏—Ç–µ —Ç–µ—Ö–Ω–∏–∫—É");
        return;
    }
    
    if (machine->getStatus() == MachineStatus::InRepair) {
        QMessageBox::information(this, "–û—Ç–ø—Ä–∞–≤–∫–∞ –≤ —Ä–µ–º–æ–Ω—Ç", "–¢–µ—Ö–Ω–∏–∫–∞ —É–∂–µ –≤ —Ä–µ–º–æ–Ω—Ç–µ");
        return;
    }
    
    if (machine->getStatus() == MachineStatus::Decommissioned) {
        QMessageBox::warning(this, "–û—Ç–ø—Ä–∞–≤–∫–∞ –≤ —Ä–µ–º–æ–Ω—Ç", "–°–ø–∏—Å–∞–Ω–Ω—É—é —Ç–µ—Ö–Ω–∏–∫—É –Ω–µ–ª—å–∑—è –æ—Ç–ø—Ä–∞–≤–∏—Ç—å –≤ —Ä–µ–º–æ–Ω—Ç");
        return;
    }
    
    // –û–±–Ω–æ–≤–ª—è–µ–º —Å—Ç–∞—Ç—É—Å
    machine->setStatus(MachineStatus::InRepair);
    if (machine->getStatus() == MachineStatus::OnSite) {
        machine->setCurrentProject("");
        machine->setAssignedDate(QDate());
    }
    
    if (FleetDatabase::instance().updateMachine(machine)) {
        m_tableModel->loadData();
        updateStatusBar();
        onTableSelectionChanged();
        QMessageBox::information(this, "–û—Ç–ø—Ä–∞–≤–∫–∞ –≤ —Ä–µ–º–æ–Ω—Ç", 
                               QString("–¢–µ—Ö–Ω–∏–∫–∞ \"%1\" –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω–∞ –≤ —Ä–µ–º–æ–Ω—Ç").arg(machine->getName()));
    } else {
        QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –æ–±–Ω–æ–≤–∏—Ç—å —Å—Ç–∞—Ç—É—Å —Ç–µ—Ö–Ω–∏–∫–∏");
    }
}

void MainWindow::onAbout()
{
    QMessageBox::about(this, "–û –ø—Ä–æ–≥—Ä–∞–º–º–µ",
                      "<h3>–ü–∞—Ä–∫ —Ç–µ—Ö–Ω–∏–∫–∏</h3>"
                      "<p>–í–µ—Ä—Å–∏—è 1.0</p>"
                      "<p>–ü—Ä–æ–≥—Ä–∞–º–º–∞ –¥–ª—è —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è –ø–∞—Ä–∫–æ–º —Å—Ç—Ä–æ–∏—Ç–µ–ª—å–Ω–æ–π —Ç–µ—Ö–Ω–∏–∫–∏</p>"
                      "<p><b>–†–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫:</b> –í—è—á–µ—Å–ª–∞–≤</p>"
                      "<p><b>–¢–µ—Ö–Ω–æ–ª–æ–≥–∏–∏:</b> Qt6, C++20, SQLite</p>"
                      "<p>–ö—É—Ä—Å–æ–≤–∞—è —Ä–∞–±–æ—Ç–∞ –ø–æ –¥–∏—Å—Ü–∏–ø–ª–∏–Ω–µ \"–ü—Ä–æ–≥—Ä–∞–º–º–∏—Ä–æ–≤–∞–Ω–∏–µ\"</p>");
}

void MainWindow::onExit()
{
    close();
}

void MainWindow::onTableSelectionChanged()
{
    auto machine = getSelectedMachine();
    updateDetailsPanel(machine);
}

void MainWindow::onStatusFilterChanged(int index)
{
    m_tableModel->setStatusFilter(index);
    updateStatusBar();
}

void MainWindow::updateDetailsPanel(MachinePtr machine)
{
    if (!machine) {
        m_detailsName->setText("‚Äî");
        m_detailsType->setText("‚Äî");
        m_detailsSerial->setText("‚Äî");
        m_detailsYear->setText("‚Äî");
        m_detailsStatus->setText("‚Äî");
        m_detailsCost->setText("‚Äî");
        m_detailsProject->setText("‚Äî");
        m_detailsAssignedDate->setText("‚Äî");
        return;
    }
    
    m_detailsName->setText(machine->getName());
    m_detailsType->setText(machine->getType());
    m_detailsSerial->setText(machine->getSerialNumber());
    m_detailsYear->setText(QString::number(machine->getYearOfManufacture()));
    
    // –°—Ç–∞—Ç—É—Å —Å —Ü–≤–µ—Ç–æ–≤—ã–º –∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏–µ–º
    QString statusText = Machine::statusToString(machine->getStatus());
    QString statusColor;
    switch (machine->getStatus()) {
        case MachineStatus::Available:
            statusColor = "#4CAF50"; break;
        case MachineStatus::OnSite:
            statusColor = "#2196F3"; break;
        case MachineStatus::InRepair:
            statusColor = "#FF9800"; break;
        case MachineStatus::Decommissioned:
            statusColor = "#F44336"; break;
    }
    m_detailsStatus->setText(QString("<span style='color: %1; font-weight: bold;'>%2</span>")
                            .arg(statusColor, statusText));
    
    m_detailsCost->setText(QString("%1 ‚ÇΩ").arg(machine->getCost(), 0, 'f', 0));
    m_detailsProject->setText(machine->getCurrentProject().isEmpty() ? "‚Äî" : machine->getCurrentProject());
    m_detailsAssignedDate->setText(machine->getAssignedDate().isValid() ?
                                  machine->getAssignedDate().toString("dd.MM.yyyy") : "‚Äî");
}

void MainWindow::updateStatusBar()
{
    auto stats = FleetDatabase::instance().getStatistics();
    QString statusText = QString("–í—Å–µ–≥–æ —Ç–µ—Ö–Ω–∏–∫–∏: %1  |  –í—ã–±—Ä–∞–Ω–æ: %2 –∏–∑ %3")
                        .arg(stats.total)
                        .arg(1) // TODO: –ü–æ–ª—É—á–∏—Ç—å —Ä–µ–∞–ª—å–Ω–æ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –≤—ã–±—Ä–∞–Ω–Ω—ã—Ö
                        .arg(24); // TODO: –ü–æ–ª—É—á–∏—Ç—å —Ä–µ–∞–ª—å–Ω–æ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç—Ñ–∏–ª—å—Ç—Ä–æ–≤–∞–Ω–Ω—ã—Ö
    
    // –î–æ–±–∞–≤–ª—è–µ–º —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫—É –ø–æ —Å—Ç–∞—Ç—É—Å–∞–º
    statusText += QString("  |  –°–≤–æ–±–æ–¥–Ω–∞: %1  |  –ù–∞ –æ–±—ä–µ–∫—Ç–∞—Ö: %2  |  –í —Ä–µ–º–æ–Ω—Ç–µ: %3")
                 .arg(stats.available)
                 .arg(stats.onSite)
                 .arg(stats.inRepair);
    
    ui->statusbar->showMessage(statusText);
}

MachinePtr MainWindow::getSelectedMachine()
{
    QModelIndexList selection = m_tableView->selectionModel()->selectedRows();
    if (selection.isEmpty()) {
        return nullptr;
    }
    
    int row = selection.first().row();
    return m_tableModel->getMachine(row);
}
